### PR TITLE
Add RVSDG tree printer pass

### PR DESCRIPTION
**Pull request by phate**
This pass enables the printing of the RVSDG tree between optimizations.

Intended usage: `jlm-opt --RvsdgTreePrinter --DeadNodeElimination --RvsdgTreePrinter -s /tmp input.ll`

This will write the files `input-rvsdgTree-0` and `input-rvsdgTree-1` to `/tmp`, containing the RVSDG tree from before applying dead node elimination and from after dead node elimination, respectively.

The plan is to extend this pass with support for annotations to the RVSDG tree. Moreover, the pass can also be generalized to print not just RVSDG tree, but also the ASCII version of the RVSDG or other debug output.

### CODE REVIEW

File: jlm/llvm/opt/RvsdgTreePrinter.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#include <jlm/llvm/ir/RvsdgModule.hpp>
+#include <jlm/llvm/opt/RvsdgTreePrinter.hpp>
+#include <jlm/util/Statistics.hpp>
+
+#include <fstream>
+
+namespace jlm::llvm
+{
+
+class RvsdgTreePrinter::Statistics final : public util::Statistics
```
Comment (phate) [inline jlm/llvm/opt/RvsdgTreePrinter.cpp:15]: FIXME: Add unit tests

File: jlm/tooling/CommandGraphGenerator.cpp [hunk 2]
```diff
@@ -126,9 +127,10 @@ JlcCommandGraphGenerator::GenerateCommandGraph(const JlcCommandLineOptions & com
 
     if (compilation.RequiresOptimization())
     {
+      util::filepath tempDirectory(std::filesystem::temp_directory_path());
       auto clangCommand = util::AssertedCast<ClangCommand>(&lastNode->GetCommand());
       auto statisticsFilePath = util::StatisticsCollectorSettings::CreateUniqueStatisticsFile(
-          util::filepath(std::filesystem::temp_directory_path()),
+          util::filepath(tempDirectory),
```
Comment (haved) [inline jlm/tooling/CommandGraphGenerator.cpp:133]: Do you need to wrap this in an additional `util::filepath`? `tempDirectory` already has that type.

**Review by haved — APPROVED** [review]
Looks good, just a couple of comments. If you are still adding some tests I can review again when they are done

File: jlm/llvm/opt/RvsdgTreePrinter.cpp (hunk 1 above)
Comment (phate) [reply jlm/llvm/opt/RvsdgTreePrinter.cpp:15]: Fixed.

File: jlm/llvm/opt/RvsdgTreePrinter.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+uint64_t
+RvsdgTreePrinter::GetOutputFileNameCounter(const RvsdgModule & rvsdgModule)
+{
+  static std::unordered_map<const RvsdgModule *, uint64_t> RvsdgModuleCounterMap_;
+
+  if (RvsdgModuleCounterMap_.find(&rvsdgModule) == RvsdgModuleCounterMap_.end())
+  {
+    RvsdgModuleCounterMap_[&rvsdgModule] = 0;
```
[the remark this answers is not in the document]
Comment (phate) [reply jlm/llvm/opt/RvsdgTreePrinter.cpp]: Fixed. Not my proudest moment...

File: jlm/tooling/CommandGraphGenerator.cpp (hunk 2 above)
Comment (phate) [reply jlm/tooling/CommandGraphGenerator.cpp]: Fixed.

**Comment by phate** [comment]
@haved Gentle reminder.

**Review by haved — APPROVED** [review]

**Pull request merged**